Source/Particles/PhysicalParticleContainer.cpp

```diff
@@ -2074,3 +2110,50 @@ PhysicalParticleContainer::buildIonizationMask (const amrex::MFIter& mfi, const
         }
     );
 }
+
+#ifdef WARPX_QED
+
+bool PhysicalParticleContainer::has_quantum_sync()
+{
+    return do_qed_quantum_sync;
+}
+
+bool PhysicalParticleContainer::has_breit_wheeler()
+{
+    return do_qed_breit_wheeler;
+}
+
+void
+PhysicalParticleContainer::
+set_breit_wheeler_engine_ptr(std::shared_ptr<BreitWheelerEngine> ptr)
+{
+    shr_ptr_bw_engine = ptr;
+}
+
+void
+PhysicalParticleContainer::
+set_quantum_sync_engine_ptr(std::shared_ptr<QuantumSynchrotronEngine> ptr)
+{
+    shr_ptr_qs_engine = ptr;
+}
+
+// A function to initialize the Tau component according to the QS engine
+void PhysicalParticleContainer::InitTauQuantumSync()
+{
+    BL_PROFILE("PhysicalParticleContainer::InitTauQuantumSync");
+    //Get functor
+    auto get_opt = shr_ptr_qs_engine->build_optical_depth_functor();
+    //Looping over all the particles
+    int num_levels = finestLevel() + 1;
+    for (int lev=0; lev < num_levels; ++lev)
+        for (WarpXParIter pti(*this, lev); pti.isValid(); ++pti){
+            auto taus = pti.GetAttribs(particle_comps["tau"]).dataPtr();
+            amrex::ParallelFor(
+                pti.numParticles(),
+                [=] AMREX_GPU_DEVICE (long i) {
+                    taus[i] = get_opt();
```

Comment:
When running a simulation with  a moving window (typically done when using boosted frame), particles are continuously injected from one side of the box. This is done by calling `AddPlasma` at each iteration, but `InitTauQuantumSync` would be called only once at the initialization of the simulation. With the current implementation, the particles starting inside the box at t=0 would have a proper `tau` initialization, but the particles that are continuously injected wouldn't (even from the same particle container).

To avoid this issue, we should put the `tau` initialization into `AddPlasma`. Would you have time to do it? I'd say you could mimic what is done for ionization (grep `do_field_ionization` in `AddPlasma`, note that member variables have to be copied to temporary variables to avoid the whole `PhysicalParticleContainer` class to be captured by valued and copied to the GPU, as done for `int loc_ionization_initial_level = ionization_initial_level;` for instance). Let me know if I can help.

Reply:
Sorry, I didn't know that. I've moved tau initialization for both photons and leptons in AddPlasma (which is part of ```PhysicalParticleContainer```).
It seems to be working correctly...